File: usage/experiments/showcase/scenario_extensions/scenarios.py
"""Scenario providers for the dedicated extension showcase."""

import random

from theseo_anysearch.experiments.custom_scenarios import ScenarioContext, ScenarioResult


def adjacent_goal_python(context: ScenarioContext) -> ScenarioResult:
    """Place a goal in one selectable voxel adjacent to the grid center."""
    center_value = (context.grid_size + 1) // 2
    start = (center_value, center_value, center_value)
    if context.scope == "evaluation":
        seed_base = int(context.parameters["seed_base"])
        index = (context.seed - seed_base) % len(context.action_offsets)
    else:
        index = random.Random(context.seed).randrange(len(context.action_offsets))
    offset = context.action_offsets[index]
    goal = tuple(start[axis] + offset[axis] for axis in range(3))
    return ScenarioResult(
        start=start,
        goal=goal,
        scenario_id=f"adjacent-{index:02d}",
        metadata={"direction_index": index, "offset": offset},
    )

File: usage/experiments/showcase/scenario_extensions/extension/src/scenarios.rs
use anysearch_extension::{anysearch_scenario, ScenarioContext, ScenarioResult};

#[anysearch_scenario]
pub fn adjacent_goal_rust(context: &ScenarioContext) -> ScenarioResult {
    let center = i32::from(context.grid_size + 1) / 2;
    let seed_base = context
        .parameters
        .get("seed_base")
        .and_then(|value| value.as_u64())
        .unwrap_or(0);
    let index = if context.scope == "evaluation" {
        context.seed.saturating_sub(seed_base) as usize % context.action_offsets.len()
    } else {
        context.seed as usize % context.action_offsets.len()
    };
    let offset = context.action_offsets[index];
    ScenarioResult::goal(
        [center, center, center],
        [center + offset[0], center + offset[1], center + offset[2]],
        format!("adjacent-{index:02}"),
    )
}
